kalite/knowledgemap/templates/knowledgemap/knowledgemap.html

```diff
@@ -0,0 +1,616 @@
+
+  var user_exercises_json;
+  var url = setGetParamDict("/api/knowledgemap/KnowledgeMapExerciselog/", {
+    user: statusModel.get("user_id"), 
+    completion_timestamp__gte: '2011-10-20',   //because the topic.json contain repetitive object under same parent, which cause problem for displaying knowledge map
+    completion_timestamp__lte: '2017-12-22'     //use this time filter to avoid encounter repetitive objects
+  }); 
+
+  doRequest(url)
+          .success(function(user_json) {
+            user_exercises_json = user_json;
+
+  //create the hashtatble 
+  var user_hashtable = {};
+  for (var key in user_exercises_json["objects"]) {
+    user_hashtable[user_exercises_json["objects"][key].exercise_id] = user_exercises_json["objects"][key];
+  }
+
+  // color-order: math_blue/sifi_scarlet/econ_orange/artist_red/cyber_green/test_purple/partner_emerald/college_pink
+  var color_scheme_array = ["#1c758a","#94424f","#c78d46","#cf5044","#699c52","#644172","#49a88f","#e89cb9"];
+
+  var subject_list = [];
+
+  var width = window.innerWidth,
+    height = window.innerHeight,
+    root;
+
+  var force = d3.layout.force()
+    // // .theta(.0001)
+    // .alpha(0.1)
+    .size([width, height]);
+    // .on("tick", tick);
+
+  $( ".kalite-navbar" ).after(function() {
+    return "<kmap></kmap>";
+  });
+
+  var svg = d3.select("kmap").append("svg")
+    .attr("width", "100%")
+    .attr("height", window.innerHeight - 180)
+    .call(d3.behavior.zoom().scaleExtent([0.01, 30]).on("zoom", zoomHandler))
+    .append("g");
+
+  var label_translate;
+  var label_scale;
+  function zoomHandler() {
+    label_translate = d3.event.translate;
+    label_scale = d3.event.scale;
+    svg.attr("transform", "translate(" + label_translate + ")scale(" + label_scale + ")");
+  }
+
+  var link = svg.selectAll(".link"),
+      node = svg.selectAll(".node");
+
+  var initialized =false;
+
+  var subj_btn_top = 70;
+  var subj_btn_left = 30;
+
+  d3.json(sprintf(TOPIC_DATA_URLS.Topics, {channel_name: "khan"}), function(error, json) {
+      root = json;
+
+      for(var c in root.children){
+        subject_list[c] = root.children[c].slug;
+      }
+      update();
+  });
+
+  var loading = svg.append("text")
+      .attr("x", width / 2)
+      .attr("y", height / 2)
+      .attr("dy", ".35em")
+      .style("text-anchor", "middle")
+      .text("Simulating. One moment please…");
+
+  var n_tick = 1;
+  var links_table = {}; //for link look up
+  var nodes;
+  var unfold_scale_factor = 1;
+  var scale_factor = 1; //for initial map size
+
+  function update() {
+
+    force.linkDistance(function(d){ 
+      if(d.target.kind !== "Topic"){
+        return  d.source.children.length * 1 + d.source.title.length * 1.5; 
+      }else{
+        return  d.source.children.length * 4 + d.target.title.length * 5; 
+      }
+    })
+    .linkStrength(function(d){
+      if(d.target.kind !== "Topic"){
+        return 1;
+      }else if(d.source.kind == "Subject"){
```

Comment:
I guess because of js asynchronous nature, the node does have a kind of "Subject". but I'm moving this code after initial(), where some nodes get their kinds assigned to "Subject".